flasc/analysis/expected_power_analysis.py

```diff
@@ -0,0 +1,757 @@
+        #  combining the expected power ratios)
+        farm_power_0 = df_sub[f"weighted_farm_power_{uplift_pair[0]}"].sum()
+        farm_power_1 = df_sub[f"weighted_farm_power_{uplift_pair[1]}"].sum()
+        total_expected_power_ratio = farm_power_1 / farm_power_0
+
+        # Compute the variance of the expected power ratio per bin
+        # This is equation 4.22
+        df_sub = df_sub.with_columns(
+            expected_power_ratio_var=(
+                (
+                    pl.col(f"pow_farm_var_{uplift_pair[0]}") * pl.col("expected_power_ratio") ** 2
+                    + pl.col(f"pow_farm_var_{uplift_pair[1]}")
+                )
+                / pl.col(f"pow_farm_{uplift_pair[0]}") ** 2
+            )
+        )
+
+        # Compute the weighted variance of the expected power ratio
+        # This is equation 4.29 pre-summation
+        df_sub = df_sub.with_columns(
+            weighted_expected_power_ratio_var=(
+                (
+                    pl.col(f"pow_farm_var_{uplift_pair[1]}") * pl.col("weight") ** 2
+                    + pl.col(f"pow_farm_var_{uplift_pair[0]}")
+                    * total_expected_power_ratio**2
+                    * pl.col("weight") ** 2
+                )
+                / farm_power_0**2
+            )
+        )
+
+        with pl.Config(tbl_cols=-1):
+            print(df_sub)
+
+        # The total uplift is the sum of the weighted expected power ratio and the weighted
+        #  variance of the expected power ratio
+        result_dict = {}
+        result_dict["energy_uplift_ctr"] = total_expected_power_ratio
+        result_dict["energy_uplift_var"] = df_sub["weighted_expected_power_ratio_var"].sum()
+
+        # Add the confidence intervals
+        z_value = norm.ppf(percentiles[1] / 100.0)
+        result_dict["energy_uplift_lb"] = result_dict["energy_uplift_ctr"] - z_value * np.sqrt(
+            result_dict["energy_uplift_var"]
+        )
+        result_dict["energy_uplift_ub"] = result_dict["energy_uplift_ctr"] + z_value * np.sqrt(
+            result_dict["energy_uplift_var"]
+        )
+
+        result_dict["df"] = df_sub
+
+        uplift_results[uplift_name] = result_dict
+
+    return uplift_results
+
+
+def total_uplift_expected_power(
+    a_in: AnalysisInput,
+    uplift_pairs: List[Tuple[str, str]],
+    uplift_names: List[str],
+    test_turbines: List[int],
+    wd_turbines: List[int] = None,
+    ws_turbines: List[int] = None,
+    use_predefined_wd: bool = False,
+    use_predefined_ws: bool = False,
+    wd_step: float = 2.0,
+    wd_min: float = 0.0,
+    wd_max: float = 360.0,
+    ws_step: float = 1.0,
+    ws_min: float = 0.0,
+    ws_max: float = 50.0,
+    bin_cols_in: List[str] = ["wd_bin", "ws_bin"],
+    weight_by: str = "min",  # min or sum
+    df_freq: pd.DataFrame = None,
+    use_standard_error: bool = True,
+    N: int = 1,
+    percentiles: List[float] = [2.5, 97.5],
+    remove_all_nulls_wd_ws: bool = False,
```

Comment:
@ejsimley should I go ahead and clear remove_all_nulls_wd_ws as it is also too much of an edge case?

Reply:
Yes, I say remove it to simplify the function, since it probably is an edge case.

Reply:
Done!